src/gt4py/storage/allocators.py

```diff
@@ -26,36 +26,62 @@
 from gt4py._core import definitions as core_defs
 from gt4py.eve import extended_typing as xtyping
 from gt4py.eve.extended_typing import (
+    TYPE_CHECKING,
     Any,
     Generic,
     NewType,
     Optional,
     Protocol,
     Sequence,
     Tuple,
-    TypeAlias,
     TypeGuard,
     TypeVar,
-    TYPE_CHECKING,
+    cast,
 )
 
+
 try:
     import cupy as cp
 except ImportError:
     cp = None
 
 
 _ScalarT = TypeVar("_ScalarT", bound=core_defs.Scalar)
+
+
+class _NDArrayLike(Protocol):
+    def __array__(self, dtype: Optional[np.dtype] = None) -> np.ndarray:
+        ...
+
+    def __getitem__(self, item: core_defs.SliceLike) -> np.ndarray:
+        ...
+
+
+class _NDArrayLikeArrayInterface(xtyping.ArrayInterface, _NDArrayLike):
+    ...
+
+
+class _NDArrayLikeCUDAArrayInterface(xtyping.CUDAArrayInterface, _NDArrayLike):
+    ...
+
+
+class _NDArrayLikeDLPackBuffer(xtyping.DLPackBuffer, _NDArrayLike):
+    ...
+
+
 _NDBufferT = TypeVar(
```

Comment:
I don't fully understand how this is supposed to be used:
- doesn't this express that it's one of the 3 things, but aren't most implementations DLPackBuffer + one of the other 2?
- I added the NDArrayLike protocol as a hack to fix some typing issues, but what did you have in mind, here?

Reply:
Yes, you're right, this is not fully correct. I think a better solution could be to define `NDBuffer` as a simple union of buffer types `Union[ArrayInterface, CUDAArrayInterface, DLPackBuffer]` and then change the _NDBufferT typevar to represent the exact array type of the tensorized view ( `NDArrayLikeT`). Alternatively, the same _NDArrayLikeT could be use in both members, which would be more precise but slightly less informative for developers.